# DSDBAL ASSIGNMENT NO. 3

Descriptive Statistics - Measures of Central Tendency and variability
Perform the following operations on any open-source dataset (e.g., data.csv)
1. Provide summary statistics (mean, median, minimum, maximum, standard deviation) for a
dataset (age, income etc.) with numeric variables grouped by one of the qualitative
(categorical) variable. For example, if your categorical variable is age groups and
quantitative variable is income, then provide summary statistics of income grouped by the
age groups. Create a list that contains a numeric value for each response to the categoriri
va
r3able.
2. Write a Python program to display some basic statistical details like percentile, mean,
standard deviation etc. of the species of ‘Iris-setosa’, ‘Iris-versicolor’ and ‘Iris- versicolor’ of
iris.csv dataset.

## 1. Importing all the required Python Libraries.

In [580]:
import numpy as np
import pandas as pd

## 2. Loading the Dataset into pandas’ data frame.

In [581]:
df=pd.read_csv('nba.csv')

In [582]:
df.head()

Name            Team  Number Position   Age Height  Weight  \
0  Avery Bradley  Boston Celtics     0.0       PG  25.0    6-2   180.0   
1    Jae Crowder  Boston Celtics    99.0       SF  25.0    6-6   235.0   
2   John Holland  Boston Celtics    30.0       SG  27.0    6-5   205.0   
3    R.J. Hunter  Boston Celtics    28.0       SG  22.0    6-5   185.0   
4  Jonas Jerebko  Boston Celtics     8.0       PF  29.0   6-10   231.0   

             College     Salary  
0              Texas  7730337.0  
1          Marquette  6796117.0  
2  Boston University        NaN  
3      Georgia State  1148640.0  
4                NaN  5000000.0

Datatypes like Number ,Age and Weight are in Float that can be converted to Int datatype as all have 
Height has (a-b) style that can be rewritten as 6.2 

In [583]:
df.shape

(458, 9)

The dataset has a total of 458 rows and 9 attributes 

## 3.Handling Missing Values

In [584]:
df.isnull().sum()

Name         1
Team         1
Number       1
Position     1
Age          1
Height       1
Weight       1
College     85
Salary      12
dtype: int64

In [585]:
for col in df.columns:
    if df[col].dtype=="float64":  # Check if column is float type
        col_mean = df[col].mean()  # Calculate mean
        df[col] = df[col].fillna(col_mean)  
df.isnull().sum()

Name         1
Team         1
Number       0
Position     1
Age          0
Height       1
Weight       0
College     85
Salary       0
dtype: int64

In [586]:
for col in df.columns:
    if df[col].dtype=='object':
        df[col]=df[col].fillna(df[col].mode()[0])

In [587]:
df.isnull().sum()

Name        0
Team        0
Number      0
Position    0
Age         0
Height      0
Weight      0
College     0
Salary      0
dtype: int64

## 4.Applying Proper Type Conversions

In [588]:
columns_to_convert = ["Number", "Age", "Weight","Salary"]  # List of columns to convert

for col in columns_to_convert:
    df[col]=df[col].astype(int)

In [589]:
df.dtypes

Name        object
Team        object
Number       int32
Position    object
Age          int32
Height      object
Weight       int32
College     object
Salary       int32
dtype: object

In [590]:
for col in df.columns:
    if df[col].dtype=='object':
        df[col]=df[col].astype('string')
df.dtypes

Name        string[python]
Team        string[python]
Number               int32
Position    string[python]
Age                  int32
Height      string[python]
Weight               int32
College     string[python]
Salary               int32
dtype: object

### Height attribute is in string type as 6-9,6-11,therefore converting it to float 

In [591]:
df['Height'] = df['Height'].fillna(df['Height'].mode()[0])

In [592]:
def convert_height(height):
    height = str(height)  # Convert to string if it's not already
    return float(height.replace('-', '.'))

df['Height']=df['Height'].apply(convert_height)
df.dtypes

Name        string[python]
Team        string[python]
Number               int32
Position    string[python]
Age                  int32
Height             float64
Weight               int32
College     string[python]
Salary               int32
dtype: object

In [593]:
df.dtypes

Name        string[python]
Team        string[python]
Number               int32
Position    string[python]
Age                  int32
Height             float64
Weight               int32
College     string[python]
Salary               int32
dtype: object

In [594]:
df.head()

Name            Team  Number Position  Age  Height  Weight  \
0  Avery Bradley  Boston Celtics       0       PG   25     6.2     180   
1    Jae Crowder  Boston Celtics      99       SF   25     6.6     235   
2   John Holland  Boston Celtics      30       SG   27     6.5     205   
3    R.J. Hunter  Boston Celtics      28       SG   22     6.5     185   
4  Jonas Jerebko  Boston Celtics       8       PF   29     6.1     231   

             College   Salary  
0              Texas  7730337  
1          Marquette  6796117  
2  Boston University  4842684  
3      Georgia State  1148640  
4           Kentucky  5000000

## 5. Central Tendencies - Measures

### 5.1.1. Mean (Built-in method)

In [595]:
for col in df.columns:
    if df[col].dtype=='int'or df[col].dtype=='float':
        print(f" Mean value of {col}: {df[col].mean()}")
        # print(f" Median value of {col}: {df[col].median()}")
        print("-"*40)


 Mean value of Number: 17.676855895196507
----------------------------------------
 Mean value of Age: 26.936681222707424
----------------------------------------
 Mean value of Height: 6.517969432314411
----------------------------------------
 Mean value of Weight: 221.52183406113537
----------------------------------------
 Mean value of Salary: 4842684.102620088
----------------------------------------


### 5.1.2. Mean (User-defined method)

In [596]:
mean={}
for col in df.columns:
    if df[col].dtype in ['int','float64']:
        total=0
        count=0
        for value in df[col]:
            if pd.notna(value):
                total=total+value
                count=count+1
        means[col]=total/count if count!=0 else None

for i,j in means.items():
    print(i," : " ,j)
    print("-"*40)

Number  :  17.676855895196507
----------------------------------------
Age  :  26.936681222707424
----------------------------------------
Height  :  6.517969432314407
----------------------------------------
Weight  :  221.52183406113537
----------------------------------------
Salary  :  4842684.102620088
----------------------------------------


### 5.2.1. Median (Built-in method)

In [597]:
for col in df.columns:
    if df[col].dtype=='int'or df[col].dtype=='float':
        print(f" Median value of {col}: {df[col].median()}")
        print("-"*40)

 Median value of Number: 13.0
----------------------------------------
 Median value of Age: 26.0
----------------------------------------
 Median value of Height: 6.6
----------------------------------------
 Median value of Weight: 220.0
----------------------------------------
 Median value of Salary: 2880600.0
----------------------------------------


### 5.2.2. Median (User-defined method)

In [598]:
medians={}

for col in df.columns:
    if df[col].dtype in ['int','float64']:
        values=sorted([x for x in df[col]])
        n=len(values)
        if n==0:
            medians[col]=None
        elif n%2==1:
            medians[col]=values[n//2]
        else:
            medians[col]=(values[n//2-1]+values[n//2])/2

for i,j in medians.items():
    print(i," : " ,j)
    print("-"*40)

Number  :  13.0
----------------------------------------
Age  :  26.0
----------------------------------------
Height  :  6.6
----------------------------------------
Weight  :  220.0
----------------------------------------
Salary  :  2880600.0
----------------------------------------


### 5.3.1. Mode (Built-in method)

In [599]:
for col in df.columns:
    # if df[col].dtype=='string':
    print(f" Mode value of {col}: {df[col].mode()[0]}")
    print("-"*40)

 Mode value of Name: Aaron Brooks
----------------------------------------
 Mode value of Team: New Orleans Pelicans
----------------------------------------
 Mode value of Number: 5
----------------------------------------
 Mode value of Position: SG
----------------------------------------
 Mode value of Age: 24
----------------------------------------
 Mode value of Height: 6.1
----------------------------------------
 Mode value of Weight: 220
----------------------------------------
 Mode value of College: Kentucky
----------------------------------------
 Mode value of Salary: 947276
----------------------------------------


### 5.3.2. Mode (User-defined method)

In [600]:
modes = {}
from collections import Counter

for col in df.columns:
    values = [x for x in df[col] if pd.notna(x)]  
    if values:  
        most_common = Counter(values).most_common()  
        max_freq = most_common[0][1]  
        mode_values = [key for key, val in most_common if val == max_freq]
        modes[col] = mode_values[0] if len(mode_values) == 1 else mode_values
    else:
        modes[col] = None 

for i,j in modes.items():
    print(i," : " ,j)
    print("-"*40)

Name  :  Aaron Brooks
----------------------------------------
Team  :  New Orleans Pelicans
----------------------------------------
Number  :  5
----------------------------------------
Position  :  SG
----------------------------------------
Age  :  24
----------------------------------------
Height  :  6.1
----------------------------------------
Weight  :  220
----------------------------------------
College  :  Kentucky
----------------------------------------
Salary  :  947276
----------------------------------------


### 5.4.1 Maximum value (Built-in method)

In [601]:
# print(df['Height'].max())
for col in df.columns:
    if df[col].dtype in ['int','float']:
        print(f'Maximum value in {col} is {df[col].max()}')
        print("-"*40)

Maximum value in Number is 99
----------------------------------------
Maximum value in Age is 40
----------------------------------------
Maximum value in Height is 7.3
----------------------------------------
Maximum value in Weight is 307
----------------------------------------
Maximum value in Salary is 25000000
----------------------------------------


### 5.4.2 Maximum value (User-defined method)

In [602]:
maximum={}
maxval=0

for col in df.columns:
    if df[col].dtype in ['int','float']:
        values=[x for x in df[col]]
        if values:
            maxval=values[0]
            for i in values:
                if i>maxval:
                    maxval=i
            maximum[col]=maxval
        else:
            maximum[col]=None  

for i,j in maximum.items():
    print(i," : " ,j)
    print("-"*40)

Number  :  99
----------------------------------------
Age  :  40
----------------------------------------
Height  :  7.3
----------------------------------------
Weight  :  307
----------------------------------------
Salary  :  25000000
----------------------------------------


### 5.5.1 Minimum value (Built-in method)

In [603]:
for col in df.columns:
    if df[col].dtype in ['int','float']:
        print(f'Minimum value in {col} is {df[col].min()}')
        print("-"*40)

Minimum value in Number is 0
----------------------------------------
Minimum value in Age is 19
----------------------------------------
Minimum value in Height is 5.11
----------------------------------------
Minimum value in Weight is 161
----------------------------------------
Minimum value in Salary is 30888
----------------------------------------


### 5.5.2 Minimum value ( User-defined method)

In [604]:
minimum={}
minval=0

for col in df.columns:
    if df[col].dtype in ['int','float']:
        values=[x for x in df[col]]
        if values:
            minval=values[0]
            for i in values:
                if i<minval:
                    minval=i
            minimum[col]=minval
        else:
            minimum[col]=None  

for i,j in minimum.items():
    print(i," : " ,j)
    print("-"*40)

Number  :  0
----------------------------------------
Age  :  19
----------------------------------------
Height  :  5.11
----------------------------------------
Weight  :  161
----------------------------------------
Salary  :  30888
----------------------------------------


### 5.6.1. Standard Deviation (Built-in method)

In [605]:
for col in df.columns:
    if df[col].dtype in ['int','float']:
        print(f"Standard Deviation of {col}: {df[col].std()}")

Standard Deviation of Number: 15.94864396617882
Standard Deviation of Age: 4.399414063696201
Standard Deviation of Height: 0.35227429750526434
Standard Deviation of Weight: 26.339488773302435
Standard Deviation of Salary: 5160125.694003381


### 5.6.2. Standard Deviation (User-defined method)

In [606]:
std_dev_values = {}

for col in df.columns:
    if df[col].dtype in ['int', 'float64']:
        values = [x for x in df[col] if pd.notna(x)] 
        n = len(values)
        if n > 1:  
            mean_value = sum(values) / n 
            squared_diffs = [(x - mean_value) ** 2 for x in values]
            variance = sum(squared_diffs) / (n - 1)
            std_dev = variance ** 0.5  
            std_dev_values[col] = std_dev  
        else:
            std_dev_values[col] = None 
for i,j in std_dev_values.items():
    print(i," : " ,j)
    print("-"*40)

Number  :  15.948643966178828
----------------------------------------
Age  :  4.399414063696205
----------------------------------------
Height  :  0.3522742975052651
----------------------------------------
Weight  :  26.33948877330243
----------------------------------------
Salary  :  5160125.694003384
----------------------------------------


### 5.7.1. Variance

In [607]:
for col in df.columns:
    if df[col].dtype in ['int','float']:
        print(f"Variance of {col}: {df[col].var()}")

Variance of Number: 254.35924435993206
Variance of Age: 19.354844103847928
Variance of Height: 0.12409718068282748
Variance of Weight: 693.768668838925
Variance of Salary: 26626897177913.875


### 5.8 Range 

In [608]:
for col in df.columns:
    if df[col].dtype in ['int','float']:
        print(f'Range of {col} is {(df[col].max())-(df[col].min())}')
        print("-"*40)

Range of Number is 99
----------------------------------------
Range of Age is 21
----------------------------------------
Range of Height is 2.1899999999999995
----------------------------------------
Range of Weight is 146
----------------------------------------
Range of Salary is 24969112
----------------------------------------


### 5.9 Quartiles

In [609]:
for col in df.columns:
    if df[col].dtype in ['int','float']:
        q1=np.percentile(df[col],25)
        q2=np.percentile(df[col],50)
        q3=np.percentile(df[col],75)
        print(f" For {col}  Q1: {q1} , Q2: {q2} ,Q3: {q3}")

 For Number  Q1: 5.0 , Q2: 13.0 ,Q3: 25.0
 For Age  Q1: 24.0 , Q2: 26.0 ,Q3: 30.0
 For Height  Q1: 6.11 , Q2: 6.6 ,Q3: 6.8
 For Weight  Q1: 200.0 , Q2: 220.0 ,Q3: 240.0
 For Salary  Q1: 1100602.0 , Q2: 2880600.0 ,Q3: 6323553.0


In [610]:
df.groupby("Salary")["Age"].apply(list)


Salary
30888           [23]
55722       [25, 27]
83397           [23]
111196          [25]
111444      [29, 25]
              ...   
22192730        [32]
22359364        [30]
22875000        [32]
22970500        [31]
25000000        [37]
Name: Age, Length: 310, dtype: object

### Grouping "Age" column by  values of "Salary" 

In [611]:
bins= [19,25,31,36,40]
labels = ['19-24','25-30','31-35','36-40']
df['AgeGroup'] = pd.cut(df['Age'], bins=bins,labels=labels)

In [612]:
df['AgeGroup'].value_counts()

AgeGroup
19-24    197
25-30    190
31-35     56
36-40     13
Name: count, dtype: int64

In [613]:
df.groupby('AgeGroup',observed=False)['Salary'].mean()

AgeGroup
19-24    3.041840e+06
25-30    6.626023e+06
31-35    5.113016e+06
36-40    5.351744e+06
Name: Salary, dtype: float64

In [614]:
df.groupby('AgeGroup',observed=False)['Salary'].median()

AgeGroup
19-24    1662360.0
25-30    4842684.0
31-35    3646250.0
36-40    4088019.0
Name: Salary, dtype: float64

In [615]:
df.groupby('AgeGroup',observed=False)['Salary'].describe()

count          mean           std       min         25%        50%  \
AgeGroup                                                                       
19-24     197.0  3.041840e+06  3.552554e+06   30888.0   947276.00  1662360.0   
25-30     190.0  6.626023e+06  5.811580e+06   55722.0  1534210.50  4842684.0   
31-35      56.0  5.113016e+06  5.283815e+06  200600.0  1323709.25  3646250.0   
36-40      13.0  5.351744e+06  6.508388e+06  222888.0   947726.00  4088019.0   

                 75%         max  
AgeGroup                          
19-24      3533333.0  16407501.0  
25-30     10126209.0  22970500.0  
31-35      6350000.0  22875000.0  
36-40      5250000.0  25000000.0

In [616]:
summary = df.groupby("Age")["Salary"].agg(["mean", "median", "min", "max", "std"])

In [617]:
summary.head()

mean     median      min       max           std
Age                                                          
19   1.930440e+06  1930440.0  1733040   2127840  2.791658e+05
20   2.725791e+06  2481720.0   525093   5703600  1.510913e+06
21   2.067380e+06  1584480.0   525093   5758680  1.412350e+06
22   2.357963e+06  1793880.0   525093   6331404  1.517378e+06
23   2.171719e+06  1201440.0    30888  16000000  2.728808e+06

In [618]:

df = pd.DataFrame(df)
numerical_data = df2[['Age', 'Weight', 'Salary']]
variance = numerical_data.var()
std_deviation = numerical_data.std()
range_values = numerical_data.max() - numerical_data.min()
quartiles = numerical_data.quantile([0.25, 0.5, 0.75])

print(f"Range:\n{range_values}\n")
print(f"Variance:\n{variance}\n")
print(f"Standard Deviation:\n{std_deviation}\n")
print(f"Quartiles (25%, 50%, 75%):\n{quartiles}")

Range:
Age             21
Weight         146
Salary    24969112
dtype: int32

Variance:
Age       1.935484e+01
Weight    6.937687e+02
Salary    2.662690e+13
dtype: float64

Standard Deviation:
Age       4.399414e+00
Weight    2.633949e+01
Salary    5.160126e+06
dtype: float64

Quartiles (25%, 50%, 75%):
       Age  Weight     Salary
0.25  24.0   200.0  1100602.0
0.50  26.0   220.0  2880600.0
0.75  30.0   240.0  6323553.0


### IRIS Dataset

Write a Python program to display some basic statistical details like percentile,
mean, standard deviation etc. of the species of ‘Iris-setosa’, ‘Iris-versicolor’ and ‘Iris-virginica’ of
iris.csv dataset.

In [619]:
df=pd.read_csv('iris.csv')

In [620]:
df.head()

Id  SepalLengthCm  SepalWidthCm  PetalLengthCm  PetalWidthCm      Species
0   1            5.1           3.5            1.4           0.2  Iris-setosa
1   2            4.9           3.0            1.4           0.2  Iris-setosa
2   3            4.7           3.2            1.3           0.2  Iris-setosa
3   4            4.6           3.1            1.5           0.2  Iris-setosa
4   5            5.0           3.6            1.4           0.2  Iris-setosa

In [621]:
df['Species'].value_counts()

Species
Iris-setosa        50
Iris-versicolor    50
Iris-virginica     50
Name: count, dtype: int64

In [622]:
df.shape

(150, 6)

In [623]:
df.isnull().sum()

Id               0
SepalLengthCm    0
SepalWidthCm     0
PetalLengthCm    0
PetalWidthCm     0
Species          0
dtype: int64

#### No Null value exists in Iris dataset 

#### Thus, Introducing Null Values in columns to Handle Missing values 

In [624]:
def introduce_null_values(df,maxnull=1):  #by default1 incase not passed 
    for col in df.columns:
        null_count=np.random.randint(1,maxnull+1) #
        
        #randomly choosing the indices on which the null values are to be set 
        indices=np.random.choice(df.index,replace=False,size=null_count)
        df.loc[indices,col]=np.nan

    return df
df=introduce_null_values(df,12)
df.isnull().sum()

Id               6
SepalLengthCm    2
SepalWidthCm     3
PetalLengthCm    4
PetalWidthCm     6
Species          8
dtype: int64

###  Handling Missing Values (Numerical and Categorical)

In [625]:
df.dtypes

Id               float64
SepalLengthCm    float64
SepalWidthCm     float64
PetalLengthCm    float64
PetalWidthCm     float64
Species           object
dtype: object

In [626]:
for col in df.columns:
    if df[col].dtype=="float64":
        df[col]=df[col].fillna(df[col].mean())

df['Species']=df['Species'].astype('string')
df['Species']=df['Species'].fillna(df['Species'].mode()[0])
df.isnull().sum()

Id               0
SepalLengthCm    0
SepalWidthCm     0
PetalLengthCm    0
PetalWidthCm     0
Species          0
dtype: int64

In [630]:
df_new = df.groupby('Species')[['SepalLengthCm', 'SepalWidthCm', 'PetalLengthCm', 'PetalWidthCm']].agg(
    {'mean','median', 'std'})
df_new.head()

SepalLengthCm                  SepalWidthCm                   \
                         mean median       std         mean median       std   
Species                                                                        
Iris-setosa          5.133929   5.05  0.513907     3.372413    3.4  0.387447   
Iris-versicolor      5.913485   5.90  0.508746     2.760747    2.8  0.321094   
Iris-virginica       6.587953   6.50  0.663002     2.971385    3.0  0.325862   

                PetalLengthCm                  PetalWidthCm                   
                         mean median       std         mean median       std  
Species                                                                       
Iris-setosa          1.877825    1.5  1.154875     0.457366    0.2  0.534548  
Iris-versicolor      4.245031    4.3  0.488899     1.327660    1.3  0.202916  
Iris-virginica       5.511877    5.5  0.619558     1.970479    2.0  0.337481

In [631]:
numeric_cols = ['SepalLengthCm', 'SepalWidthCm', 'PetalLengthCm', 'PetalWidthCm']

# Group by 'Species' and calculate statistics
summary_stats = df.groupby("Species")[numeric_cols].describe(percentiles=[0.25, 0.5, 0.75])
summary_stats.head().transpose()


Species              Iris-setosa  Iris-versicolor  Iris-virginica
SepalLengthCm count    56.000000        47.000000       47.000000
              mean      5.133929         5.913485        6.587953
              std       0.513907         0.508746        0.663002
              min       4.300000         4.900000        4.900000
              25%       4.800000         5.600000        6.150000
              50%       5.050000         5.900000        6.500000
              75%       5.400000         6.250000        7.000000
              max       6.600000         7.000000        7.900000
SepalWidthCm  count    56.000000        47.000000       47.000000
              mean      3.372413         2.760747        2.971385
              std       0.387447         0.321094        0.325862
              min       2.300000         2.000000        2.200000
              25%       3.100000         2.500000        2.800000
              50%       3.400000         2.800000        3.000000
              75%       3.600000         3.000000        3.100000
              max       4.400000         3.400000        3.800000
PetalLengthCm count    56.000000        47.000000       47.000000
              mean      1.877825         4.245031        5.511877
              std       1.154875         0.488899        0.619558
              min       1.000000         3.000000        3.758219
              25%       1.400000         3.950000        5.100000
              50%       1.500000         4.300000        5.500000
              75%       1.600000         4.600000        5.850000
              max       6.000000         5.100000        6.900000
PetalWidthCm  count    56.000000        47.000000       47.000000
              mean      0.457366         1.327660        1.970479
              std       0.534548         0.202916        0.337481
              min       0.100000         1.000000        1.204167
              25%       0.200000         1.200000        1.800000
              50%       0.200000         1.300000        2.000000
              75%       0.400000         1.500000        2.300000
              max       2.500000         1.800000        2.500000